uat/custom-components/client-java-paho/src/main/java/com/aws/greengrass/testing/mqtt5/client/paho/MqttConnectionImpl.java

```diff
@@ -266,15 +272,37 @@ private IMqttAsyncClient createAsyncClient(MqttLib.ConnectionParams connectionPa
     private void disconnectAndClose(long timeout, int reasonCode, List<Mqtt5Properties> userProperties)
             throws org.eclipse.paho.mqttv5.common.MqttException {
         MqttProperties properties = new MqttProperties();
+
         if (userProperties != null && !userProperties.isEmpty()) {
             properties.setUserProperties(convertToUserProperties(userProperties));
             userProperties.forEach(p -> logger.atInfo()
                     .log("Disconnect MQTT userProperties: {}, {}", p.getKey(), p.getValue()));
         }
+
         try {
-            client.disconnectForcibly(QUIESCE_TIMEOUT, timeout, reasonCode, properties);
+            if (isConnected.compareAndSet(true, false)) {
+                client.disconnectForcibly(QUIESCE_TIMEOUT, timeout, reasonCode, properties);
+            } else {
+                logger.atWarn().log("DISCONNECT was not sent on the dead connection");
+            }
+
+            final long deadline = System.nanoTime() + timeout * 1_000_000_000;
+
+            long remaining = deadline - System.nanoTime();
+            if (remaining < MIN_SHUTDOWN_NS) {
+                remaining = MIN_SHUTDOWN_NS;
+            }
+
+            executorService.shutdown();
+            if (!executorService.awaitTermination(remaining, TimeUnit.NANOSECONDS)) {
+                executorService.shutdownNow();
+            }
+        } catch (InterruptedException e) {
+            executorService.shutdownNow();
+            Thread.currentThread().interrupt();
         } finally {
             client.close();
+            executorService.shutdown();
```

Comment:
shutdown twice ?
It was already called in line 296.

Also when client.close() will throw an exception line 305 will never called